### PR TITLE
add warning to `Iterators.filter` about assumptions on predicate

### PR DESCRIPTION
**Pull request by adienes**
addresses #50440

### CODE REVIEW

**Comment by adienes** [comment]
bump

File: base/iterators.jl [hunk 1]
```diff
@@ -491,6 +491,11 @@ invocation of `filter`. Calls to `flt` will be made when iterating over the
 returned iterable object. These calls are not cached and repeated calls will be
 made when reiterating.
 
+!!! warning
+    The implementation may assume that `flt` is pure, so please use caution
+    to ensure correctness when the predicate values may be non-deterministic,
+    have side-effects, or be dependent on the order of iteration.
```
Comment (timholy) [inline base/iterators.jl:497]: Can we be more specific about the types of problems that might be encountered. "Please use caution" might be OK for a sign that starts with "Ice forms on bridge, please use caution" but in this context it's not obviously actionable advice.

Comment (adienes) [reply base/iterators.jl]: woops, I was on that `map!` branch and accidentally force pushed it here. one min

Comment (adienes) [reply base/iterators.jl]: ok, reworded

**Comment by timholy** [comment]
Very nice, thanks!

**Pull request merged**